### ISSUE TITLE
Prevent Imgui from scaling image when window resizes

### ISSUE THREAD

**Issue by LightShade12** [ocornut/imgui#5941]
Version/Branch of Dear ImGui:
Version: 1.89 WIP
Branch: docking

Back-end/Renderer/Compiler/OS:
Back-ends: imgui_impl_opengl3.cpp + imgui_impl_glfw.cpp
Operating System: Windows 10

My Issue/Question:
How could I go about preventing an ImGui window from scaling a single image in it when the window resizes?
I'm rendering a texture that changes its actual resolution when container window is resized. This worked fine when I was using GLFW window as it doesn't scale textures automatically when window resizes. But when trying to do the same in ImGui, my image gets twice the scaling.
Note that I want the scaling I have implemented since it saves performance, I just want to turn off image scaling from ImGui side.

Initial scale:
![image](https://user-images.githubusercontent.com/102753608/204818945-21cf3a38-6d39-4360-a84d-0066c782ad46.png)
Result when increasing window size:
![image](https://user-images.githubusercontent.com/102753608/204819152-02d8a92d-1af3-4d21-8d61-e70895b6ab0a.png)
(Due to extreme scaling, image pixels are visible)
Result when decreasing window size:
![image](https://user-images.githubusercontent.com/102753608/204819419-395777d9-18b8-4a34-9ec4-8bf23ec4568f.png)
(image gets reduced twice as much)

Currently I'm using this code to add an image to ImGui window:
```
ImVec2 pos = ImGui::GetCursorScreenPos();
    ImVec2 ws = ImGui::GetContentRegionAvail();
    ImGui::GetWindowDrawList()->AddImage(
        (void*)texID,
        ImVec2(ImGui::GetCursorScreenPos()),
        ImVec2(ImGui::GetCursorScreenPos().x + ws.x,
            ImGui::GetCursorScreenPos().y + ws.y), ImVec2(0, 1), ImVec2(1, 0));
```

**Comment by ocornut** [comment]
The code looks correct, I suspect you have an unrelated issue with your rendering to texture.

**Comment by LightShade12** [comment]
> The code looks correct, I suspect you have an unrelated issue with your rendering to texture.

Thanks for the fast response. The code works fine for static image resolutions as they need to be scaled according to the window. Since my texture has variable resolution, I want to prevent image scaling from ImGui side (The texture maintains its actual pixel size(w.r.t screen pixels) regardless of whether the window changes size or not). Is this possible?

**Comment by ocornut** [comment]
Dear ImGui draws the image in the size you give it. You gave it `ImGui::GetContentRegionAvail()` as output size. It's all under your control.

**Comment by ocornut** [comment]
I suspect it's either an issue with your render to texture either an issue with you not specifying the right texture coordinates (see e.g. https://github.com/ocornut/imgui/wiki/Image-Loading-and-Displaying-Examples#About-texture-coordinates about Texture Coordinates).

**Comment by GamingMinds-DanielC** [comment]
Problems like this can be analyzed pretty good in [RenderDoc](https://renderdoc.org/).

Launch your application through RenderDoc, then it will hook into all graphics calls. You can make captures (F12 per default) and see every single draw call in detail. And you can view the exact contents of your texture and all the details of the geometry used to draw it. Very useful to pinpoint the problem.

**Comment by LightShade12** [comment]
> Problems like this can be analyzed pretty good in [RenderDoc](https://renderdoc.org/).
> 
> Launch your application through RenderDoc, then it will hook into all graphics calls. You can make captures (F12 per default) and see every single draw call in detail. And you can view the exact contents of your texture and all the details of the geometry used to draw it. Very useful to pinpoint the problem.

Thank you. I looked at the website. Since its kinda complicated, it might take a few days for me to get used to it. I'm actually rendering on the cpu so I'm wondering if renderdoc will capture my custom cpu code or not. It seems it mainly works with graphics apis rendering.

**Comment by ocornut** [comment]
> I don't understand, this is just another texture created as any valid opengl texture is created, it should scale and repeat itself, but doesn't it do that?

This depends on your texture sampler settings, looks like yours is set to clamping on both axises.

**Comment by GamingMinds-DanielC** [comment]
> I'm actually rendering on the cpu so I'm wondering if renderdoc will capture my custom cpu code or not.

You stated an OpenGL backend above. So if you are rendering the contents of your texture on the CPU, RenderDoc will not capture that. But it will capture the resource update and the GL calls to draw ImGui. So you will be able to see exactly what the content of your texture is when you give it to ImGui for display.

And don't worry about apparent complexity. RenderDoc can look intimidating at first glance, but when you use it and just take a look around, it is rather straightforward. Plus there are Youtube tutorials on how to use it, they should get you started quickly.

**Comment by ocornut** [comment]
@LightShade12 think of it this way: all AddImage() does is to create a textured rectangle (made of 2 triangles) with the edge positions and UV you gave to it. In other words, your issue is not a Dear ImGui issue.

If I look at this:
![image](https://user-images.githubusercontent.com/8225057/204859525-50cf9ddb-3819-458b-80fc-14ff88f86d0f.png)

Your OpenGL texture very likely contains the pixel inside the red rectangle:
![image](https://user-images.githubusercontent.com/8225057/204860344-7dd1bbb9-e5bf-427b-b14a-13701f5f8806.png)
In other word, part of your texture contains undrawn or invalid data.
This may be perfectly fine as long as you are not trying to display this section, which you are here.
If you are only rendering to part of your texture, you need to provide matching UV coordinates to only display the drawn part.

The upper-most and right-most section with the long line are UV coordinates not in the 0.0..1.0 range (<0 or >1) and your texture sample is set to CLAMP for those axis, so the last pixel is repeated.
See https://www.cs.uaf.edu/2006/fall/cs381/lecture/10_10_texture.html
You may make your texture repeat with:
```cpp
// bind texture, then:
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
```

**Comment by LightShade12** [comment]
> @LightShade12 think of it this way: all AddImage() does is to create a textured rectangle (made of 2 triangles) with the edge positions and UV you gave to it. In other words, your issue is not a Dear ImGui issue.
> 
> If I look at this: ![image](https://user-images.githubusercontent.com/8225057/204859525-50cf9ddb-3819-458b-80fc-14ff88f86d0f.png)
> 
> Your OpenGL texture very likely contains the pixel inside the red rectangle: ![image](https://user-images.githubusercontent.com/8225057/204860344-7dd1bbb9-e5bf-427b-b14a-13701f5f8806.png) In other word, part of your texture contains undrawn or invalid data. This may be perfectly fine as long as you are not trying to display this section, which you are here. If you are only rendering to part of your texture, you need to provide matching UV coordinates to only display the drawn part.
> 
> The upper-most and right-most section with the long line are UV coordinates not in the 0.0..1.0 range (<0 or >1) and your texture sample is set to CLAMP for those axis, so the last pixel is repeated. See https://www.cs.uaf.edu/2006/fall/cs381/lecture/10_10_texture.html You may make your texture repeat with:
> 
> ```c++
> // bind texture, then:
> glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
> glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
> ```

Unfortunately this doesn't seem to work either :
![image](https://user-images.githubusercontent.com/102753608/204936937-8cee1333-c94d-4af4-8f11-7d8c3b6bc10c.png)
though I do think maybe its not something ImGui is doing, but rather the texture itself...

**Comment by LightShade12** [comment]
> > I'm actually rendering on the cpu so I'm wondering if renderdoc will capture my custom cpu code or not.
> 
> You stated an OpenGL backend above. So if you are rendering the contents of your texture on the CPU, RenderDoc will not capture that. But it will capture the resource update and the GL calls to draw ImGui. So you will be able to see exactly what the content of your texture is when you give it to ImGui for display.
> 
> And don't worry about apparent complexity. RenderDoc can look intimidating at first glance, but when you use it and just take a look around, it is rather straightforward. Plus there are Youtube tutorials on how to use it, they should get you started quickly.

I have downloaded Renderdoc and when I press f12 it captures only the empty GLFW window, how do I go about capturing ImGui windows only?

**Comment by LightShade12** [comment]
ITS FIXED!
![image](https://user-images.githubusercontent.com/102753608/204953607-d2a64c67-a173-4510-b5be-a85f4a37c0f2.png)
After checking render docs, I confirmed that it really was the texture itself and not ImGui. 
turns out because of the way I was scaling the texture, window and viewport was mixed up so I just had to separate these dimensions(I was reusing dimensions since on GLFW windowres=texres). It was causing incomplete rendering into the GLFW window when I scaled the viewport.
I had to separate window res and tex res for my renderer and then I could change just tex res by scaling viewport while the Image was rendered in a full GLFW window.
